STMReco/src/PlotSTMWaveformDigis_module.cc

```diff
@@ -0,0 +1,70 @@
+//
+// Analyzer module to plot STM waveforms
+//
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Core/EDAnalyzer.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "cetlib_except/exception.h"
+#include "fhiclcpp/types/Atom.h"
+#include "canvas/Utilities/InputTag.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+#include "art_root_io/TFileService.h"
+
+#include <utility>
+#include <numeric>
+// root
+#include "TH1F.h"
+#include "TF1.h"
+#include "TGraph.h"
+
+#include "Offline/RecoDataProducts/inc/STMWaveformDigi.hh"
+
+namespace mu2e {
+
+  class PlotSTMWaveformDigis : public art::EDAnalyzer {
+    public:
+      using Name=fhicl::Name;
+      using Comment=fhicl::Comment;
+      struct Config {
+        fhicl::Atom<art::InputTag> stmWaveformDigisTag{ Name("stmWaveformDigisTag"), Comment("InputTag for STMWaveformDigiCollection")};
+      };
+      using Parameters = art::EDAnalyzer::Table<Config>;
+      explicit PlotSTMWaveformDigis(const Parameters& conf);
+
+    private:
+    void analyze(const art::Event& e) override;
+
+    art::InputTag _stmWaveformDigisTag;
+  };
+
+  PlotSTMWaveformDigis::PlotSTMWaveformDigis(const Parameters& config )  :
+    art::EDAnalyzer{config},
+    _stmWaveformDigisTag(config().stmWaveformDigisTag())
+  {
+    consumes<STMWaveformDigiCollection>(_stmWaveformDigisTag);
+  }
+
+  void PlotSTMWaveformDigis::analyze(const art::Event& event) {
+
+    art::ServiceHandle<art::TFileService> tfs;
+    auto waveformsHandle = event.getValidHandle<STMWaveformDigiCollection>(_stmWaveformDigisTag);
+
+    std::stringstream histname;
+    int count = 0;
+    for (const auto& waveform : *waveformsHandle) {
+      histname.str("");
+      histname << "evt" << event.event() << "_waveform" << count;
+      TH1F* _hWaveform = tfs->make<TH1F>(histname.str().c_str(), "", waveform.adcs().size(),0,waveform.adcs().size());
+      int i_bin = 1;
+      for (const auto& adc : waveform.adcs()) {
```

Comment:
Another FYI.  I have no idea what the compiler is going to do with this.   My rule of thumb is that if an object is smaller than a pointer, work with it by value, not by reference or pointer.  Since an adc value is smaller than 8 bytes, I would write this without the reference.  I bet that in this case the whole discussion is moot - the code is simple enough that the optimizer will probably do the obvious thing.

Reply:
I've decided to keep this as it is. Hopefully the compiler is optimizing it. If we see performance issues, I can change it